cura/Machines/MaterialNode.py

```diff
@@ -1,25 +1,119 @@
-# Copyright (c) 2018 Ultimaker B.V.
+# Copyright (c) 2019 Ultimaker B.V.
 # Cura is released under the terms of the LGPLv3 or higher.
-from typing import Optional, Dict, Any
-from collections import OrderedDict
-from .ContainerNode import ContainerNode
 
+from typing import Any, TYPE_CHECKING
 
+from UM.Logger import Logger
+from UM.Settings.ContainerRegistry import ContainerRegistry
+from UM.Settings.Interfaces import ContainerInterface
+from UM.Signal import Signal
+from cura.Machines.ContainerNode import ContainerNode
+from cura.Machines.QualityNode import QualityNode
+
+if TYPE_CHECKING:
+    from typing import Dict
+    from cura.Machines.VariantNode import VariantNode
+
+##  Represents a material in the container tree.
 #
-# A MaterialNode is a node in the material lookup tree/map/table. It contains 2 (extra) fields:
-#  - material_map: a one-to-one map of "material_root_id" to material_node.
-#  - children_map: the key-value map for child nodes of this node. This is used in a lookup tree.
-#
-#
+#   Its subcontainers are quality profiles.
 class MaterialNode(ContainerNode):
-    __slots__ = ("material_map", "children_map")
+    def __init__(self, container_id, variant: "VariantNode") -> None:
+        super().__init__(container_id)
+        self.variant = variant
+        self.qualities = {}  # type: Dict[str, QualityNode] # Mapping container IDs to quality profiles.
+        self.materialChanged = Signal()  # Triggered when the material is removed or its metadata is updated.
+
+        container_registry = ContainerRegistry.getInstance()
+        my_metadata = container_registry.findContainersMetadata(id = container_id)[0]
+        self.base_file = my_metadata["base_file"]
+        self.material_type = my_metadata["material"]
+        self.guid = my_metadata["GUID"]
+        self._loadAll()
+        container_registry.containerRemoved.connect(self._onRemoved)
+        container_registry.containerMetaDataChanged.connect(self._onMetadataChanged)
+
+    ##  Finds the preferred quality for this printer with this material and this
+    #   variant loaded.
+    #
+    #   If the preferred quality is not available, an arbitrary quality is
+    #   returned. If there is a configuration mistake (like a typo in the
+    #   preferred quality) this returns a random available quality. If there are
+    #   no available qualities, this will return the empty quality node.
+    #   \return The node for the preferred quality, or any arbitrary quality if
+    #   there is no match.
+    def preferredQuality(self) -> QualityNode:
```

Comment:
Minor: I'd call it `getPreferredQuality()` because it's a function.

Reply:
I guess it's a style thing. I don't think that every function that returns something useful should be prefixed with `get`.